### PR TITLE
[405B] Add performance data for 405B model

### PR DESCRIPTION
**Pull request by fduwjj**
In this PR, we mostly measured the performance and loss curves for 405B model with some optimizations techniques we recently developed. We also want to log the actual peak TFLOPs used for MFU calculation for cross-validation. Also we should get device information from system rather from device name because it does not contain "NVL" or "SXM".

image

### CODE REVIEW

File: docs/performance.md [hunk 1]
```diff
@@ -19,6 +19,21 @@ Next we show the loss curves for Llama 3 8B and Llama 3 70B training with both 1
 ![image](../assets/images/llama3_loss_curves.png)
 
 
+## Llama 3.1 performance numbers
+
+We did a performance measurement on the 405B model released in [LLaMa 3.1](https://llama.meta.com/docs/model-cards-and-prompt-formats/llama3_1). Because the model now is larger, we run on 128 H100 GPUs to test both performance and loss curves. Below is the performance result of 405B model with optimizations we have developed. We do see OOM for 1D even with batch size = 1, so we only tested the 2D case.
+
+
+| Model size | Batch size | Activation checkpointing | WPS | MFU | optimizations |
```
Comment (tianyu-l) [inline docs/performance.md]:
```suggestion
| Model size | Batch size | Activation checkpointing | WPS | MFU | Optimizations |
```

File: docs/performance.md [hunk 2]
```diff
@@ -19,6 +19,21 @@ Next we show the loss curves for Llama 3 8B and Llama 3 70B training with both 1
 ![image](../assets/images/llama3_loss_curves.png)
 
 
+## Llama 3.1 performance numbers
```
Comment (tianyu-l) [inline docs/performance.md]: I suggest we move this section to top, to make things in reverse chronological order.

File: torchtitan/utils.py [hunk 3]
```diff
@@ -145,7 +148,12 @@ def get_peak_flops(device_name: str) -> int:
             return 1979e12
         elif "PCIe" in device_name:
             return 756e12
-        else:  # for SXM and other variants
+        # Sometimes, the device name is just "NVIDIA H100", we need
+        # to check the memory size to determine the peak flops.
+        # H100 NVL
```
Comment (tianyu-l) [inline torchtitan/utils.py]: what's this comment for? what's its relationship with L147 above?

File: train.py [hunk 4]
```diff
@@ -71,7 +71,8 @@ def main(job_config: JobConfig):
     utils.init_distributed(job_config)
     # initialize GPU memory monitor and get peak flops for MFU calculation
     gpu_memory_monitor = build_gpu_memory_monitor()
-    gpu_peak_flops = utils.get_peak_flops(gpu_memory_monitor.device_name)
+    gpu_peak_flops = utils.get_peak_flops(gpu_memory_monitor)
+    logger.info(f"Peak FLOPS we use for MFU: {gpu_peak_flops} ")
```
Comment (tianyu-l) [inline train.py]:
```suggestion
    logger.info(f"Peak FLOPS used for computing MFU: {gpu_peak_flops} ")
```

Comment (tianyu-l) [reply train.py]: Do we need to do some formatting of the peak flops, as I can imagine it's a very large number?

File: assets/images/llama3_1_405B_loss_curves.png
Comment (tianyu-l) [inline assets/images/llama3_1_405B_loss_curves.png:1]: looks great!

File: docs/performance.md [hunk 5]
```diff
@@ -19,6 +19,21 @@ Next we show the loss curves for Llama 3 8B and Llama 3 70B training with both 1
...
+
+We did a performance measurement on the 405B model released in [LLaMa 3.1](https://llama.meta.com/docs/model-cards-and-prompt-formats/llama3_1). Because the model now is larger, we run on 128 H100 GPUs to test both performance and loss curves. Below is the performance result of 405B model with optimizations we have developed. We do see OOM for 1D even with batch size = 1, so we only tested the 2D case.
+
+
+| Model size | Batch size | Activation checkpointing | WPS | MFU | optimizations |
+| ----- | ----- | ----- | ----- | ----- | ----- |
+| 405B | 2 | full | 118 | 37.1% | None
+| 405B | 2 | full | 177 | 27.77% | FP8
+| 405B | 2 | full | 185 | 29.03% | FP8 + async TP
```
Comment (tianyu-l) [inline docs/performance.md]: I think now they call it Float8 instead of FP8, e.g. see https://dev-discuss.pytorch.org/t/enabling-float8-all-gather-in-fsdp2/2359

Let's follow that.

**Review by tianyu-l — COMMENTED** [review]
Looks good to me, except that the peak flops section is still a bit mysterious to me.

File: torchtitan/utils.py (hunk 3 above)
Comment (tianyu-l) [reply torchtitan/utils.py]: ideally we should still include pointers to how these numbers were obtained... if possible

File: train.py (hunk 4 above)
Comment (fduwjj) [reply train.py]: sure, I can add a format for it.

File: docs/performance.md (hunk 2 above)
Comment (fduwjj) [reply docs/performance.md]: But the wording needs to change, let me do that.

File: torchtitan/utils.py (hunk 3 above)
Comment (fduwjj) [reply torchtitan/utils.py]: L147 is never hit because like I said in the comment, device name is literally "NVIDIA H100" there is no "NVL" or "SXM" in its name. So L147 is never hit in our case.

This is the screenshot from our run: 
image

You can see that the device name is "NVIDIA H100"

Comment (fduwjj) [reply torchtitan/utils.py]: The number comes from the same source you listed in L145. The only single source is L145.

File: docs/performance.md [hunk 6]
```diff
@@ -2,9 +2,25 @@ To demonstrate the effectiveness of PyTorch distributed training techniques used
 We report infra metrics achieved by [FSDP2](fsdp.md) (1D parallelism) under various configurations, and loss curves for both 1D parallelism (FSDP2) and 2D parallelism (FSDP2 + Tensor Parallel) training.
```
Comment (tianyu-l) [inline docs/performance.md]: we'll have to do rewording of the top section as well.

File: docs/performance.md [hunk 7]
```diff
@@ -1,10 +1,26 @@
...
 
 
+## Llama 3.1 performance numbers
+
+Below are the WPS (word per second, or more accurately, token per second) and MFU (model FLOPS utilization) results which torchtitan achieves on the 405B model released in [Llama 3.1](https://llama.meta.com/docs/model-cards-and-prompt-formats/llama3_1). The way we compute WPS and MFU can be found in `train.py`. Because the model now is larger, we run on 128 H100 GPUs to test both performance and loss curves. Below is the performance result of 405B model with optimizations we have developed. We do see OOM for 1D parallelism (FSDP2), so we only tested 2D parallelism (FSDP2 + Tensor Parallel).
+
+| Model size | Batch size | Activation checkpointing | WPS | MFU | Optimizations |
+| ----- | ----- | ----- | ----- | ----- | ----- |
+| 405B | 2 | full | 118 | 37.1% | None
+| 405B | 2 | full | 177 | 27.77% | Float8
```
Comment (drisspg) [inline docs/performance.md]: One other thing to note for Float8 is that we are not converting all the matmuls to Float8, since our fused attention implementation is not done in fp8 I suspect this is lower than actuality

Comment (fduwjj) [reply docs/performance.md]: Do you think we want a footnote for what you say here?

File: torchtitan/utils.py [hunk 8]
```diff
@@ -145,7 +148,13 @@ def get_peak_flops(device_name: str) -> int:
             return 1979e12
```
Comment (yifuwang) [inline torchtitan/utils.py]: Can you help fix this while you are here?

File: torchtitan/utils.py [hunk 9]
```diff
@@ -145,7 +148,13 @@ def get_peak_flops(device_name: str) -> int:
...
         elif "PCIe" in device_name:
             return 756e12
-        else:  # for SXM and other variants
+        # Sometimes, the device name is just literally "NVIDIA H100", so
+        # we won't hit the two conditions before; we then need
+        # to check the memory size of the device to determine the peak flops.
+        # H100 NVL has 94 GiB memory, source: https://www.nvidia.com/en-us/data-center/h100/
+        elif device_mem >= 94:
```
Comment (yifuwang) [inline torchtitan/utils.py]: I don't think this is reliable. There are H100 SXM variants with 96GB. Meta's H100s also have the SXM form factor despite having 96GB HBM2e.

I wonder if we should use `lspci` to determine the form factor.

File: torchtitan/utils.py (hunk 8 above)
Comment (fduwjj) [reply torchtitan/utils.py]: sure

**Review by tianyu-l — APPROVED** [review]
let's remove and delay the changes on peak flops (as it's more of a separate topic which needs further discussion) and land the rest for now.
